### PR TITLE
Better errors for logging corner cases

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/12863

### Does your PR introduce any breaking changes? If yes, please list them.

Logging a non-scalar tensor now raises an error instead of computing the `mean` internally.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### CODE REVIEW

File: pytorch_lightning/core/module.py [hunk 1]
```diff
@@ -531,6 +531,14 @@ def __check_allowed(v: Any, name: str, value: Any) -> None:
     def __to_tensor(self, value: numbers.Number) -> torch.Tensor:
         return torch.tensor(value, device=self.device)
 
+    @staticmethod
+    def __check_numel_1(value: torch.Tensor, name: str) -> None:
+        if not torch.numel(value) == 1:
```
Comment (awaelchli) [inline pytorch_lightning/core/module.py:536]: LGTM assuming extra dims get squeezed out?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/core/module.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/core/module.py:536]: They don't. But everything is working as expected in some tests I run with multi-dim single-element tensors

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
